<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Yusha Liu" />


<title>Testing differential expression on simulated bulk rna-seq data under the null (same permutation)</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">tbdata_sim</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/YushaLiu/tbdata_sim">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Testing differential expression on simulated bulk rna-seq data under the null (same permutation)</h1>
<h4 class="author">Yusha Liu</h4>
<h4 class="date">2021-2-2</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-02-03
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>tbdata_sim/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20210202code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20210202)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20210202code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20210202)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomYushaLiutbdatasimtreedd01471aca6badef511c4b281e9535560cbf424ctargetblankdd01471a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/YushaLiu/tbdata_sim/tree/dd01471aca6badef511c4b281e9535560cbf424c" target="_blank">dd01471</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomYushaLiutbdatasimtreedd01471aca6badef511c4b281e9535560cbf424ctargetblankdd01471a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/YushaLiu/tbdata_sim/tree/dd01471aca6badef511c4b281e9535560cbf424c" target="_blank">dd01471</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    output/.DS_Store

Untracked files:
    Untracked:  output/mouthwash_sim_null_summary.Rds
    Untracked:  output/mouthwash_sim_null_v2_summary.Rds
    Untracked:  output/null_output/
    Untracked:  output/null_v2_output/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/sim_null.Rmd</code>) and HTML (<code>docs/sim_null.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/YushaLiu/tbdata_sim/blob/dd01471aca6badef511c4b281e9535560cbf424c/analysis/sim_null.Rmd" target="_blank">dd01471</a>
</td>
<td>
yushaliu
</td>
<td>
2021-02-03
</td>
<td>
Update some analyses
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/YushaLiu/tbdata_sim/3066607a7b57887209431dfad1dcabf3bd4b226a/docs/sim_null.html" target="_blank">3066607</a>
</td>
<td>
yushaliu
</td>
<td>
2021-02-03
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/YushaLiu/tbdata_sim/blob/f4cf8b833534f829734c9cffe37c68995c112de8/analysis/sim_null.Rmd" target="_blank">f4cf8b8</a>
</td>
<td>
yushaliu
</td>
<td>
2021-02-03
</td>
<td>
Add the simulation results
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="simulation-procedure" class="section level2">
<h2>Simulation procedure</h2>
<p>We apply various methods to detect differential exression on bulk rna-seq data simulated under the null. More specifically, we start with a subset of the <a href="https://pubmed.ncbi.nlm.nih.gov/26586179/">tbdata</a> which includes the gene expression data of 12,728 genes from 54 samples infected by 8 different types of bacteria and one non-infected control (6 individuals per condition), measured at 4 hours post-infection. Then we randomly permute the samples with respect to the condition labels so that there no longer exists association between gene expression levels and conditions.</p>
<p>There are two different ways to randomly permute the data: 1) the same permutation to all the genes, so the correlation among genes is preserved on the permuted dataset; 2) independent and different permutation to each gene. Here we simulate null data by applying the same permutation to all the genes; we simulate 10 replicate datasets in total.</p>
</div>
<div id="methods" class="section level2">
<h2>Methods</h2>
<p>We consider the following methods to compare gene expression levels under each infected condition to the control on the simulated null data. For methods related to mash, we always assume <span class="math inline">\(\alpha = 1\)</span>, i.e., the EZ model.</p>
<p>1. limma + mash with common baseline.</p>
<p>2. sva + limma + mash with common baseline.</p>
<p>3. mash + sample-specific random effects.</p>
<p>4. mash + sample-specific random effects + ruv, with variance inflation factor constrained to be at least 1.</p>
<p>5. poisson mash + sample-specific random effects + ruv.</p>
</div>
<div id="look-at-distributions-of-z-scores-by-comparing-to-the-mean" class="section level2">
<h2>Look at distributions of Z-scores by comparing to the mean</h2>
<pre class="r"><code>library(mashr)
library(edgeR)
library(limma)

### load in the design matrix
design &lt;- readRDS(&quot;../simulations/design.Rds&quot;)
trts &lt;- colnames(design)
R &lt;- length(trts)

### plot the histograms of Z-scores across all genes and conditions
par(mfrow=c(2,2))

for(idx in 1:10){
  ### read in data
  counts &lt;- readRDS(paste0(&quot;../simulations/null_data/raw_data&quot;, idx, &quot;.Rds&quot;))
  
  ### normalize and transform the count data
  y &lt;- DGEList(counts)
  y &lt;- calcNormFactors(y)
  dat_cpm &lt;- cpm(y, log = TRUE)
  
  ### extract Bhat and Shat using limma  
  cov_of_interest = 1:ncol(design)
  lmout = limma::lmFit(object = dat_cpm, design = design)
  eout  = limma::eBayes(lmout)
  bhat = lmout$coefficients[,cov_of_interest,drop=FALSE]
  shat = lmout$stdev.unscaled[,cov_of_interest,drop=FALSE] * sqrt(eout$s2.post)
  rownames(shat) = rownames(bhat)
  
  ### compute Z-scores by comparing to the overall mean across all conditions
  mash.data = mash_set_data(bhat, shat, alpha = 1)
  mash.data.L = mash_update_data(mash.data, ref = &quot;mean&quot;)
  mash.Z = mash.data.L$Bhat/mash.data.L$Shat
  
  ### plot the histograms of Z-scores of all genes and conditions
  hist(mash.Z, xlab=&quot;Z-scores&quot;, main=paste0(&quot;Hist of Z-scores \n by comparing to the mean \n across genes and conditions for rep=&quot;, idx))
}</code></pre>
<p><img src="figure/sim_null.Rmd/unnamed-chunk-1-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/sim_null.Rmd/unnamed-chunk-1-2.png" width="672" style="display: block; margin: auto;" /><img src="figure/sim_null.Rmd/unnamed-chunk-1-3.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="result-summary" class="section level2">
<h2>Result summary</h2>
<pre class="r"><code>### limma + mash with common baseline
num_effects_1 &lt;- rep(NA, 10)

for(i in 1:10){
  mash.fit &lt;- readRDS(paste0(&quot;output/null_output/mash_fit_rep_&quot;, i, &quot;.Rds&quot;))
  idx.mash &lt;- get_significant_results(mash.fit)
  num_effects_1[i] &lt;- length(idx.mash)
}

### number of false positives
num_effects_1</code></pre>
<pre><code> [1]    6 3958    5 1856 3073  952 6158 1285 2937 4744</code></pre>
<pre class="r"><code>### false positive rates
round(num_effects_1/nrow(mash.fit$result$lfsr), 3)</code></pre>
<pre><code> [1] 0.000 0.311 0.000 0.146 0.242 0.075 0.484 0.101 0.231 0.373</code></pre>
<pre class="r"><code>### sva + limma + mash with common baseline 
num_effects_2 &lt;- rep(NA, 10)

for(i in 1:10){
  mash.fit &lt;- readRDS(paste0(&quot;output/null_output/sva_mash_fit_rep_&quot;, i, &quot;.Rds&quot;))
  idx.mash &lt;- get_significant_results(mash.fit)
  num_effects_2[i] &lt;- length(idx.mash)
}

### number of false positives
num_effects_2</code></pre>
<pre><code> [1] 7138 8527 7535 7924 6685 6515 6795 8047 7452 8521</code></pre>
<pre class="r"><code>### false positive rates
round(num_effects_2/nrow(mash.fit$result$lfsr), 3)</code></pre>
<pre><code> [1] 0.561 0.670 0.592 0.623 0.526 0.512 0.534 0.633 0.586 0.670</code></pre>
<pre class="r"><code>### mash + sample-specific random effects
num_effects_3 &lt;- rep(NA, 10)

for(i in 1:10){
  mash.fit &lt;- readRDS(paste0(&quot;output/null_output/mash_no_ruv_posterior_rep_&quot;, i, &quot;.Rds&quot;))
  num_effects_3[i] &lt;- sum(apply(mash.fit$lfsr, 1, min) &lt; 0.05)
}

### number of false positives
num_effects_3</code></pre>
<pre><code> [1]    2 5175  126 1458 2288  621 4970 1724 3031 5041</code></pre>
<pre class="r"><code>### false positive rates
round(num_effects_3/nrow(mash.fit$lfsr), 3)</code></pre>
<pre><code> [1] 0.000 0.407 0.010 0.115 0.180 0.049 0.391 0.136 0.238 0.396</code></pre>
<pre class="r"><code>### mash + sample-specific random effects + ruv
num_effects_4 &lt;- rep(NA, 10)

for(i in 1:10){
  mash.fit &lt;- readRDS(paste0(&quot;output/null_output/mash_ruv_posterior_rep_&quot;, i, &quot;.Rds&quot;))
  num_effects_4[i] &lt;- sum(apply(mash.fit$lfsr, 1, min) &lt; 0.05)
}

### number of false positives
num_effects_4</code></pre>
<pre><code> [1] 1461 5230 2976 4862 2667 3374 6815 5653 3293 4258</code></pre>
<pre class="r"><code>### false positive rates
round(num_effects_4/nrow(mash.fit$lfsr), 3)</code></pre>
<pre><code> [1] 0.115 0.411 0.234 0.382 0.210 0.265 0.536 0.444 0.259 0.335</code></pre>
<pre class="r"><code>### poisson mash + sample-specific random effects + ruv
num_effects_5 &lt;- rep(NA, 10)

for(i in 1:10){
  pois.fit &lt;- readRDS(paste0(&quot;output/null_output/pois_mash_ruv_posterior_rep_&quot;, i, &quot;.Rds&quot;))
  num_effects_5[i] &lt;- sum(apply(pois.fit$lfsr, 1, min) &lt; 0.05)
}

### number of false positives
num_effects_5</code></pre>
<pre><code> [1]  9103 10105 10702  8318  9999 10532 10673  9190  6583 10953</code></pre>
<pre class="r"><code>### false positive rates
round(num_effects_5/nrow(pois.fit$lfsr), 3)</code></pre>
<pre><code> [1] 0.716 0.794 0.841 0.654 0.786 0.828 0.839 0.722 0.518 0.861</code></pre>
</div>
<div id="look-at-estimated-weights-of-prior-covariances-in-the-mash-ruv-fit" class="section level2">
<h2>Look at estimated weights of prior covariances in the mash ruv fit</h2>
<pre class="r"><code>library(pheatmap)
library(gridExtra)

## load in the mash ruv fit 
for(i in 1:5){
  res &lt;- readRDS(paste0(&quot;../simulations/null_output/mash_ruv_fit_rep_&quot;, i, &quot;.Rds&quot;))
  
  # calculate estimated weights
  pi.mat &lt;- matrix(res$pi, ncol=length(res$wlist), byrow=TRUE)
  rownames(pi.mat) &lt;- c(names(res$Ulist), names(res$ulist))
  colnames(pi.mat) &lt;- paste0(&quot;w=&quot;, round(res$wlist, 3))

  # plot estimated weights
  pheatmap(pi.mat, cluster_rows=FALSE, cluster_cols=FALSE, fontsize_row=10, fontsize_col=10, 
           main=paste0(&quot;Estimated weights of prior covariances in mash ruv fit \n for null data rep=&quot;, i))
  
}</code></pre>
<p><img src="figure/sim_null.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/sim_null.Rmd/unnamed-chunk-7-2.png" width="672" style="display: block; margin: auto;" /><img src="figure/sim_null.Rmd/unnamed-chunk-7-3.png" width="672" style="display: block; margin: auto;" /><img src="figure/sim_null.Rmd/unnamed-chunk-7-4.png" width="672" style="display: block; margin: auto;" /><img src="figure/sim_null.Rmd/unnamed-chunk-7-5.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="look-at-pairwise-comparisons-with-control-using-mouthwash-sprop1" class="section level2">
<h2>Look at pairwise comparisons with control using MOUTHWASH (sprop=1)</h2>
<pre class="r"><code>m.out &lt;- readRDS(&quot;output/mouthwash_sim_null_summary.Rds&quot;)

## the estimate of the proportion of null genes
round(m.out$pi0,3)</code></pre>
<pre><code>       [,1]  [,2]  [,3]  [,4]  [,5]  [,6]  [,7]  [,8]
 [1,] 0.522 0.995 0.060 0.732 0.121 0.901 0.763 0.411
 [2,] 0.285 0.095 0.847 0.157 0.364 0.034 0.360 0.067
 [3,] 0.242 0.528 0.883 0.413 0.647 0.821 0.851 0.395
 [4,] 0.395 0.055 0.202 0.089 0.143 0.049 0.141 0.082
 [5,] 0.997 0.450 0.924 0.970 0.893 0.926 0.823 0.977
 [6,] 0.156 0.297 0.024 0.741 0.081 0.234 0.192 0.030
 [7,] 0.009 0.118 0.054 0.690 0.063 0.195 0.053 0.032
 [8,] 0.996 0.242 0.993 0.334 0.038 0.609 0.197 0.231
 [9,] 0.916 0.250 0.879 0.275 0.414 0.415 0.256 0.509
[10,] 0.883 0.607 0.845 0.263 0.281 0.966 0.752 0.293</code></pre>
<pre class="r"><code>## the number of DE genes by comparing each infected condition to the control (lfsr &lt; 0.05)
m.out$num_effects_1</code></pre>
<pre><code>      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
 [1,]   14    0  544   59  251    6   39   41
 [2,]  412  639   76  113   48 4275  187 2022
 [3,] 1232  291   35  176   66    5   18   64
 [4,]   81 1020  442 2745  234  250  577 1989
 [5,]    0  221   22    1   14    1    6   10
 [6,]  670   16 2581   31 1290 1137  346 2118
 [7,] 4906  821 4539   46 1902 1718 1839 4865
 [8,]    4  599    2   67 2981  437  346  183
 [9,]   16  588   71   84   46  462  905  173
[10,]   16   52    4  457  170    3    6   68</code></pre>
<pre class="r"><code>## the number of DE genes by comparing each infected condition to the control (lfsr &lt; 0.05/8)
m.out$num_effects_2</code></pre>
<pre><code>      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
 [1,]    3    0   48   20   25    1   13    6
 [2,]   95  105   29   12    3 1929   54  541
 [3,]  414  100   10   33   20    1    4   12
 [4,]   10  167  109 1100   29   16   89  660
 [5,]    0   66    6    0    6    1    2    5
 [6,]  149    1  629    7  254  492   63  438
 [7,] 2048  175 2364    9  463  654  373 2499
 [8,]    2  236    1   10  980  246   50   31
 [9,]    5  150   24    8    2  170  317   41
[10,]    7   13    2   81   20    0    1    6</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.2 (2019-12-12)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS  10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] gridExtra_2.3   pheatmap_1.0.12 edgeR_3.28.1    limma_3.42.2   
[5] mashr_0.2.21    ashr_2.2-50     workflowr_1.6.2

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.6         RColorBrewer_1.1-2 compiler_3.6.2     later_1.0.0       
 [5] git2r_0.27.1       plyr_1.8.6         tools_3.6.2        digest_0.6.27     
 [9] lifecycle_0.2.0    gtable_0.3.0       evaluate_0.14      lattice_0.20-41   
[13] rlang_0.4.10       Matrix_1.2-18      yaml_2.2.1         mvtnorm_1.1-0     
[17] xfun_0.17          invgamma_1.1       stringr_1.4.0      knitr_1.30        
[21] fs_1.4.1           locfit_1.5-9.4     rprojroot_2.0.2    grid_3.6.2        
[25] glue_1.4.2         R6_2.5.0           rmarkdown_2.3      mixsqp_0.3-43     
[29] irlba_2.3.3        rmeta_3.0          magrittr_2.0.1     whisker_0.4       
[33] scales_1.1.1       promises_1.1.0     htmltools_0.4.0    abind_1.4-5       
[37] assertthat_0.2.1   colorspace_2.0-0   httpuv_1.5.2       stringi_1.4.6     
[41] munsell_0.5.0      truncnorm_1.0-8    SQUAREM_2021.1    </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
