<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Yusha Liu" />


<title>Testing differential expression on simulated bulk rna-seq data under the null (same permutation)</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">tbdata_sim</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/YushaLiu/tbdata_sim">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Testing differential expression on simulated bulk rna-seq data under the null (same permutation)</h1>
<h4 class="author">Yusha Liu</h4>
<h4 class="date">2021-3-1</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-03-02
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>tbdata_sim/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20210202code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20210202)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20210202code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20210202)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomYushaLiutbdatasimtree6768f00ad3367d30cb785a3311938d729dbdc524targetblank6768f00a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/YushaLiu/tbdata_sim/tree/6768f00ad3367d30cb785a3311938d729dbdc524" target="_blank">6768f00</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomYushaLiutbdatasimtree6768f00ad3367d30cb785a3311938d729dbdc524targetblank6768f00a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/YushaLiu/tbdata_sim/tree/6768f00ad3367d30cb785a3311938d729dbdc524" target="_blank">6768f00</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    output/.DS_Store

Untracked files:
    Untracked:  code/code_sim_null_summary.R
    Untracked:  code/code_sim_null_v2_summary.R
    Untracked:  output/null_output/
    Untracked:  output/null_v2_output/
    Untracked:  output/sim_null_limma_mash_results.Rds
    Untracked:  output/sim_null_mash_ruv_results.Rds
    Untracked:  output/sim_null_mouthwash_K_3_results.Rds
    Untracked:  output/sim_null_mouthwash_K_3_t_lik_results.Rds
    Untracked:  output/sim_null_mouthwash_K_6_results.Rds
    Untracked:  output/sim_null_mouthwash_K_6_t_lik_results.Rds
    Untracked:  output/sim_null_pois_mash_ruv_results.Rds
    Untracked:  output/sim_null_sva_limma_mash_results.Rds
    Untracked:  output/sim_null_v2_limma_mash_results.Rds
    Untracked:  output/sim_null_v2_mash_ruv_results.Rds
    Untracked:  output/sim_null_v2_mouthwash_K_3_results.Rds
    Untracked:  output/sim_null_v2_mouthwash_K_3_t_lik_results.Rds
    Untracked:  output/sim_null_v2_pois_mash_ruv_results.Rds

Unstaged changes:
    Deleted:    analysis/sim_null_ash.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/sim_null_pairwise.Rmd</code>) and HTML (<code>docs/sim_null_pairwise.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/YushaLiu/tbdata_sim/blob/6768f00ad3367d30cb785a3311938d729dbdc524/analysis/sim_null_pairwise.Rmd" target="_blank">6768f00</a>
</td>
<td>
yushaliu
</td>
<td>
2021-03-02
</td>
<td>
update sim results
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/YushaLiu/tbdata_sim/09956f360f082380160c3957c929bb1c590dba82/docs/sim_null_pairwise.html" target="_blank">09956f3</a>
</td>
<td>
yushaliu
</td>
<td>
2021-03-02
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/YushaLiu/tbdata_sim/blob/49f553ce61548f8f6f6c6ff4a78057a4ca03850d/analysis/sim_null_pairwise.Rmd" target="_blank">49f553c</a>
</td>
<td>
yushaliu
</td>
<td>
2021-03-02
</td>
<td>
update sim results
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="simulation-procedure" class="section level2">
<h2>Simulation procedure</h2>
<p>We apply various methods to detect differential exression on bulk rna-seq data simulated under the null. More specifically, we start with a subset of the <a href="https://pubmed.ncbi.nlm.nih.gov/26586179/">tbdata</a> which includes the gene expression data of 12,728 genes from 54 samples infected by 8 different types of bacteria and one non-infected control (6 individuals per condition), measured at 4 hours post-infection. Then we randomly permute the samples with respect to the condition labels so that there no longer exists association between gene expression levels and conditions.</p>
<p>There are two different ways to randomly permute the data: 1) the same permutation to all the genes, so the correlation among genes is preserved on the permuted dataset; 2) independent and different permutation to each gene. Here we simulate null data by applying the same permutation to all the genes; we simulate 10 replicate datasets in total.</p>
</div>
<div id="methods" class="section level2">
<h2>Methods</h2>
<p>We apply ash and mouthwash to do pairwise comparisons of gene expressions between each infected condition and the control on the simulated null data.</p>
<p>We implement the following 4 versions of ash:</p>
<p>1. <span class="math inline">\(\alpha=1\)</span> (EZ model), original shat.</p>
<p>2. <span class="math inline">\(\alpha=1\)</span> (EZ model), shat calibrated based on Z-score.</p>
<p>3. <span class="math inline">\(\alpha=0\)</span> (EE model), original shat.</p>
<p>4. <span class="math inline">\(\alpha=0\)</span> (EE model), shat calibrated based on Z-score.</p>
<p>We implement the following 4 versions of mouthwash, while setting <span class="math inline">\(\alpha=1\)</span> (EZ model) in all cases:</p>
<p>1. Number of hidden confounders <span class="math inline">\(K=3\)</span>, original shat.</p>
<p>2. Number of hidden confounders <span class="math inline">\(K=3\)</span>, shat calibrated based on Z-score.</p>
<p>3. Number of hidden confounders <span class="math inline">\(K=6\)</span>, original shat.</p>
<p>4. Number of hidden confounders <span class="math inline">\(K=6\)</span>, shat calibrated based on Z-score.</p>
</div>
<div id="load-in-the-packages-and-the-shat-calibration-function" class="section level2">
<h2>Load in the packages and the shat calibration function</h2>
<pre class="r"><code>library(ashr)
library(edgeR)
library(limma)

### the function to calibrate shat under t likelihood based on Z score
adjust_by_t &lt;- function(betahat, sebetahat, df) {
  ## Check input -----------------------------------------
  assertthat::are_equal(length(betahat), length(sebetahat))
  assertthat::assert_that(length(betahat) == length(df) | length(df) == 1)
  assertthat::assert_that(df &gt; 0)
  
  ## Convert t to z --------------------------------------
  zstats &lt;- stats::qnorm(stats::pt(q = betahat / sebetahat, df = df))
  snew &lt;- c(betahat / zstats)
  
  return(snew)
}


### load in the design matrix
design &lt;- readRDS(&quot;../simulations/design.Rds&quot;)
trts &lt;- colnames(design)
R &lt;- length(trts)</code></pre>
</div>
<div id="ash-alpha1-ez-model-original-shat." class="section level2">
<h2>Ash: <span class="math inline">\(\alpha=1\)</span> (EZ model), original shat.</h2>
<pre class="r"><code>num_effects1 &lt;- matrix(NA, nrow=10, ncol=R-1)
rownames(num_effects1) &lt;- paste0(&quot;rep&quot;, 1:10)
colnames(num_effects1) &lt;- paste0(&quot;cond&quot;, 1:(R-1))

for(idx in 1:10){
  scdata &lt;- readRDS(paste0(&quot;../simulations/null_data/raw_data&quot;, idx, &quot;.Rds&quot;))
  
  ### iterate over conditions 
  for(r in 1:8){
    ### take data from one infected condition and the control
    counts &lt;- scdata[, (design[,1]==1 | design[,r+1]==1)]
    
    ### normalize and transform the count data
    y &lt;- DGEList(counts)
    y &lt;- calcNormFactors(y)
    dat_cpm &lt;- cpm(y, log = TRUE)
    
    ### construct the covariate matrix
    data.X &lt;- model.matrix(~design[(design[,1]==1 | design[,r+1]==1), r+1])
    colnames(data.X) &lt;- c(&quot;Intercept&quot;, &quot;Trt&quot;)
    
    ### compute bhat and shat (of the difference between the given condition and the negative control)
    cov_of_interest = 2
    lmout = limma::lmFit(object = dat_cpm, design = data.X)
    eout = limma::eBayes(lmout)
    bhat = lmout$coefficients[,cov_of_interest]
    shat = lmout$stdev.unscaled[,cov_of_interest] * sqrt(eout$s2.post)
    
    ### apply ash
    ash.out = ash(betahat=bhat, sebetahat=shat, mixcompdist=&quot;normal&quot;, alpha=1, lik=lik_normal())
    num_effects1[idx,r] = sum(ash.out$result$lfsr &lt; 0.05)
  }
}


### the number of significant genes by comparing each infected condition to the control (lfsr &lt; 0.05)
num_effects1</code></pre>
<pre><code>      cond1 cond2 cond3 cond4 cond5 cond6 cond7 cond8
rep1      0     0     0     0     0     0     0     0
rep2   1524     6     2     0     0  1110     6    17
rep3    266    13     0     0     0     0     0     0
rep4      3     1     1   365     0     0     0   158
rep5      0     4     0     0     0     0     0     0
rep6      3     0   105     0     0    60     0     1
rep7   3434   798  2815     3  3149   488    33  3357
rep8      0    49     0     0    30   121     4     0
rep9      0    36     0     1     0     5    11   267
rep10     0     0     0     0   489     0     0     0</code></pre>
</div>
<div id="ash-alpha1-ez-model-shat-calibrated-based-on-z-score." class="section level2">
<h2>Ash: <span class="math inline">\(\alpha=1\)</span> (EZ model), shat calibrated based on Z-score.</h2>
<pre class="r"><code>num_effects2 &lt;- matrix(NA, nrow=10, ncol=R-1)
rownames(num_effects2) &lt;- paste0(&quot;rep&quot;, 1:10)
colnames(num_effects2) &lt;- paste0(&quot;cond&quot;, 1:(R-1))

for(idx in 1:10){
  scdata &lt;- readRDS(paste0(&quot;../simulations/null_data/raw_data&quot;, idx, &quot;.Rds&quot;))
  
  ### iterate over conditions 
  for(r in 1:8){
    ### take data from one infected condition and the control
    counts &lt;- scdata[, (design[,1]==1 | design[,r+1]==1)]
    
    ### normalize and transform the count data
    y &lt;- DGEList(counts)
    y &lt;- calcNormFactors(y)
    dat_cpm &lt;- cpm(y, log = TRUE)
    
    ### construct the covariate matrix
    data.X &lt;- model.matrix(~design[(design[,1]==1 | design[,r+1]==1), r+1])
    colnames(data.X) &lt;- c(&quot;Intercept&quot;, &quot;Trt&quot;)
    
    ### compute bhat and shat (of the difference between the given condition and the negative control)
    cov_of_interest = 2
    lmout = limma::lmFit(object = dat_cpm, design = data.X)
    eout = limma::eBayes(lmout)
    bhat = lmout$coefficients[,cov_of_interest]
    shat = lmout$stdev.unscaled[,cov_of_interest] * sqrt(eout$s2.post)
    shat.adj = adjust_by_t(bhat, shat, df=ncol(dat_cpm)-2)
    
    ### apply ash
    ash.out = ash(betahat=bhat, sebetahat=shat.adj, mixcompdist=&quot;normal&quot;, alpha=1, lik=lik_normal())
    num_effects2[idx,r] = sum(ash.out$result$lfsr &lt; 0.05)
  }
}

### the number of significant genes by comparing each infected condition to the control (lfsr &lt; 0.05)
num_effects2</code></pre>
<pre><code>      cond1 cond2 cond3 cond4 cond5 cond6 cond7 cond8
rep1      0     0     0     0     0     0     0     0
rep2    214     0     0     0     0   261     0     0
rep3      2     0     0     0     0     0     0     0
rep4      0     0     0    35     0     0     0     1
rep5      0     0     0     0     0     0     0     0
rep6      0     0     0     0     0     0     0     0
rep7   1904    36  1092     0  1303     9     0  1605
rep8      0     0     0     0     0     1     0     0
rep9      0     0     0     0     0     0     0     0
rep10     0     0     0     0    10     0     0     0</code></pre>
</div>
<div id="ash-alpha0-ee-model-original-shat." class="section level2">
<h2>Ash: <span class="math inline">\(\alpha=0\)</span> (EE model), original shat.</h2>
<pre class="r"><code>num_effects3 &lt;- matrix(NA, nrow=10, ncol=R-1)
rownames(num_effects3) &lt;- paste0(&quot;rep&quot;, 1:10)
colnames(num_effects3) &lt;- paste0(&quot;cond&quot;, 1:(R-1))

for(idx in 1:10){
  scdata &lt;- readRDS(paste0(&quot;../simulations/null_data/raw_data&quot;, idx, &quot;.Rds&quot;))
  
  ### iterate over conditions 
  for(r in 1:8){
    ### take data from one infected condition and the control
    counts &lt;- scdata[, (design[,1]==1 | design[,r+1]==1)]
    
    ### normalize and transform the count data
    y &lt;- DGEList(counts)
    y &lt;- calcNormFactors(y)
    dat_cpm &lt;- cpm(y, log = TRUE)
    
    ### construct the covariate matrix
    data.X &lt;- model.matrix(~design[(design[,1]==1 | design[,r+1]==1), r+1])
    colnames(data.X) &lt;- c(&quot;Intercept&quot;, &quot;Trt&quot;)
    
    ### compute bhat and shat (for the difference between the given condition and the negative control)
    cov_of_interest = 2
    lmout = limma::lmFit(object = dat_cpm, design = data.X)
    eout = limma::eBayes(lmout)
    bhat = lmout$coefficients[,cov_of_interest]
    shat = lmout$stdev.unscaled[,cov_of_interest] * sqrt(eout$s2.post)
    
    ### apply ash
    ash.out = ash(betahat=bhat, sebetahat=shat, mixcompdist=&quot;normal&quot;, alpha=0, lik=lik_normal())
    num_effects3[idx,r] = sum(ash.out$result$lfsr &lt; 0.05)
  }
}

### the number of significant genes by comparing each infected condition to the control (lfsr &lt; 0.05)
num_effects3</code></pre>
<pre><code>      cond1 cond2 cond3 cond4 cond5 cond6 cond7 cond8
rep1      0     0     0     0     0     0     0     0
rep2    902     4     2     0     0   530     4    13
rep3    144    10     0     0     0     0     0     0
rep4      1     1     0   249     0     0     0    98
rep5      0     1     0     0     0     0     0     0
rep6      3     0    60     0     0    54     0     2
rep7   3280   170  2303     0  2735   237     3  2616
rep8      0    25     0     0    26    62     1     0
rep9      0    12     0     1     0     5     9    94
rep10     0     0     0     0   113     0     0     0</code></pre>
</div>
<div id="ash-alpha0-ee-model-shat-calibrated-based-on-z-score." class="section level2">
<h2>Ash: <span class="math inline">\(\alpha=0\)</span> (EE model), shat calibrated based on Z-score.</h2>
<pre class="r"><code>num_effects4 &lt;- matrix(NA, nrow=10, ncol=R-1)
rownames(num_effects4) &lt;- paste0(&quot;rep&quot;, 1:10)
colnames(num_effects4) &lt;- paste0(&quot;cond&quot;, 1:(R-1))

for(idx in 1:10){
  scdata &lt;- readRDS(paste0(&quot;../simulations/null_data/raw_data&quot;, idx, &quot;.Rds&quot;))
  
  ### iterate over conditions 
  for(r in 1:8){
    ### take data from one infected condition and the control
    counts &lt;- scdata[, (design[,1]==1 | design[,r+1]==1)]
    
    ### normalize and transform the count data
    y &lt;- DGEList(counts)
    y &lt;- calcNormFactors(y)
    dat_cpm &lt;- cpm(y, log = TRUE)
    
    ### construct the covariate matrix
    data.X &lt;- model.matrix(~design[(design[,1]==1 | design[,r+1]==1), r+1])
    colnames(data.X) &lt;- c(&quot;Intercept&quot;, &quot;Trt&quot;)
    
    ### compute bhat and shat (for the difference between the given condition and the negative control)
    cov_of_interest = 2
    lmout = limma::lmFit(object = dat_cpm, design = data.X)
    eout = limma::eBayes(lmout)
    bhat = lmout$coefficients[,cov_of_interest]
    shat = lmout$stdev.unscaled[,cov_of_interest] * sqrt(eout$s2.post)
    shat.adj = adjust_by_t(bhat, shat, df=ncol(dat_cpm)-2)
    
    ### apply ash
    ash.out = ash(betahat=bhat, sebetahat=shat.adj, mixcompdist=&quot;normal&quot;, alpha=0, lik=lik_normal())
    num_effects4[idx,r] = sum(ash.out$result$lfsr &lt; 0.05)
  }
}

### the number of significant genes by comparing each infected condition to the control (lfsr &lt; 0.05)
num_effects4</code></pre>
<pre><code>      cond1 cond2 cond3 cond4 cond5 cond6 cond7 cond8
rep1      0     0     0     0     0     0     0     0
rep2     41     0     0     0     0     9     0     0
rep3      0     0     0     0     0     0     0     0
rep4      0     0     0     0     0     0     0     0
rep5      0     0     0     0     0     0     0     0
rep6      0     0     0     0     0     0     0     0
rep7   1657     0   667     0   913     0     0  1072
rep8      0     0     0     0     0     0     0     0
rep9      0     0     0     0     0     0     0     0
rep10     0     0     0     0     0     0     0     0</code></pre>
</div>
<div id="mouthwash-number-of-hidden-confounders-k3-original-shat." class="section level2">
<h2>Mouthwash: number of hidden confounders <span class="math inline">\(K=3\)</span>, original shat.</h2>
<pre class="r"><code>m.out &lt;- readRDS(&quot;output/sim_null_mouthwash_K_3_results.Rds&quot;)

### the estimate of the proportion of null genes
round(m.out$pi0, 3)</code></pre>
<pre><code>      cond1 cond2 cond3 cond4 cond5 cond6 cond7 cond8
rep1  0.522 0.995 0.060 0.732 0.121 0.901 0.763 0.411
rep2  0.285 0.095 0.847 0.157 0.364 0.034 0.360 0.067
rep3  0.242 0.528 0.883 0.413 0.647 0.821 0.851 0.395
rep4  0.395 0.055 0.202 0.089 0.143 0.049 0.141 0.082
rep5  0.997 0.450 0.924 0.970 0.893 0.926 0.823 0.977
rep6  0.156 0.297 0.024 0.741 0.081 0.234 0.192 0.030
rep7  0.009 0.118 0.054 0.690 0.063 0.195 0.053 0.032
rep8  0.996 0.242 0.993 0.334 0.038 0.609 0.197 0.231
rep9  0.916 0.250 0.879 0.275 0.414 0.415 0.256 0.509
rep10 0.883 0.607 0.845 0.263 0.281 0.966 0.752 0.293</code></pre>
<pre class="r"><code>### the number of significant genes by comparing each infected condition to the control (lfsr &lt; 0.05)
m.out$num_effects</code></pre>
<pre><code>      cond1 cond2 cond3 cond4 cond5 cond6 cond7 cond8
rep1     14     0   544    59   251     6    39    41
rep2    412   639    76   113    48  4275   187  2022
rep3   1232   291    35   176    66     5    18    64
rep4     81  1020   442  2745   234   250   577  1989
rep5      0   221    22     1    14     1     6    10
rep6    670    16  2581    31  1290  1137   346  2118
rep7   4906   821  4539    46  1902  1718  1839  4865
rep8      4   599     2    67  2981   437   346   183
rep9     16   588    71    84    46   462   905   173
rep10    16    52     4   457   170     3     6    68</code></pre>
</div>
<div id="mouthwash-number-of-hidden-confounders-k3-shat-calibrated-based-on-z-score." class="section level2">
<h2>Mouthwash: number of hidden confounders <span class="math inline">\(K=3\)</span>, shat calibrated based on Z-score.</h2>
<pre class="r"><code>m.out &lt;- readRDS(&quot;output/sim_null_mouthwash_K_3_t_lik_results.Rds&quot;)

### the estimate of the proportion of null genes
round(m.out$pi0, 3)</code></pre>
<pre><code>      cond1 cond2 cond3 cond4 cond5 cond6 cond7 cond8
rep1  0.984 0.997 0.075 0.978 0.208 0.992 0.952 0.654
rep2  0.365 0.075 0.962 0.236 0.508 0.011 0.443 0.027
rep3  0.075 0.436 0.989 0.246 0.918 0.996 0.985 0.716
rep4  0.842 0.039 0.095 0.021 0.184 0.089 0.078 0.031
rep5  0.998 0.515 0.995 0.995 0.995 0.994 0.988 0.995
rep6  0.074 0.940 0.023 0.947 0.040 0.073 0.114 0.022
rep7  0.007 0.108 0.012 0.924 0.052 0.070 0.042 0.010
rep8  0.998 0.114 0.997 0.420 0.017 0.834 0.164 0.238
rep9  0.995 0.102 0.959 0.320 0.522 0.226 0.074 0.895
rep10 0.988 0.834 0.991 0.099 0.711 0.995 0.989 0.372</code></pre>
<pre class="r"><code>### the number of significant genes by comparing each infected condition to the control (lfsr &lt; 0.05)
m.out$num_effects</code></pre>
<pre><code>      cond1 cond2 cond3 cond4 cond5 cond6 cond7 cond8
rep1      1     0     6     0     5     0     0     0
rep2     29    37     2     0     0  2863     8   812
rep3    546     8     0     2     1     0     0     0
rep4      2   162    55  1880     1     2    59  1098
rep5      0     7     0     0     0     0     0     0
rep6    106     0   836     0   344   665    14   596
rep7   2857    87  3470     2   432   939   400  3642
rep8      0   130     0     0  1449   158    10     2
rep9      1    87     0     0     0    68   490     3
rep10     0     0     0    43     0     0     0     0</code></pre>
</div>
<div id="mouthwash-number-of-hidden-confounders-k6-original-shat." class="section level2">
<h2>Mouthwash: number of hidden confounders <span class="math inline">\(K=6\)</span>, original shat.</h2>
<pre class="r"><code>m.out &lt;- readRDS(&quot;output/sim_null_mouthwash_K_6_results.Rds&quot;)

### the estimate of the proportion of null genes
round(m.out$pi0, 3)</code></pre>
<pre><code>      cond1 cond2 cond3 cond4 cond5 cond6 cond7 cond8
rep1  0.417 0.623 0.535 0.141 0.495 0.336 0.335 0.059
rep2  0.120 0.259 0.337 0.485 0.202 0.133 0.191 0.132
rep3  0.036 0.129 0.082 0.446 0.278 0.473 0.418 0.101
rep4  0.268 0.487 0.149 0.062 0.273 0.522 0.485 0.205
rep5  0.868 0.383 0.540 0.787 0.561 0.803 0.675 0.513
rep6  0.073 0.515 0.081 0.508 0.109 0.172 0.105 0.208
rep7  0.033 0.222 0.029 0.561 0.078 0.093 0.197 0.038
rep8  0.913 0.558 0.838 0.620 0.144 0.262 0.424 0.189
rep9  0.196 0.077 0.225 0.407 0.399 0.418 0.117 0.124
rep10 0.422 0.383 0.135 0.197 0.595 0.679 0.719 0.172</code></pre>
<pre class="r"><code>### the number of significant genes by comparing each infected condition to the control (lfsr &lt; 0.05)
m.out$num_effects</code></pre>
<pre><code>      cond1 cond2 cond3 cond4 cond5 cond6 cond7 cond8
rep1    403   226   368  1859   463   432   908  2001
rep2   2232  1030   816   353   495  2944  1113  2409
rep3   5725  2319  1892   733   725   428   493   847
rep4    715   505  1529  4406   492   269   515  1505
rep5     32   904   386    47   410    62   209   254
rep6   3094   286  2784   340  2303  2627  1681  1124
rep7   4164  1526  5847   623  2279  3232  1555  5912
rep8     58   607    40   248  2032  1513   547  1123
rep9    972  2580  1093   332   657  1271  2447  2463
rep10   468   851   860  1703   407   166   168   937</code></pre>
</div>
<div id="mouthwash-number-of-hidden-confounders-k6-shat-calibrated-based-on-z-score." class="section level2">
<h2>Mouthwash: number of hidden confounders <span class="math inline">\(K=6\)</span>, shat calibrated based on Z-score.</h2>
<pre class="r"><code>m.out &lt;- readRDS(&quot;output/sim_null_mouthwash_K_6_t_lik_results.Rds&quot;)

### the estimate of the proportion of null genes
round(m.out$pi0, 3)</code></pre>
<pre><code>      cond1 cond2 cond3 cond4 cond5 cond6 cond7 cond8
rep1  0.847 0.902 0.882 0.069 0.817 0.231 0.096 0.041
rep2  0.245 0.123 0.191 0.668 0.143 0.048 0.185 0.039
rep3  0.007 0.035 0.044 0.366 0.335 0.495 0.204 0.138
rep4  0.642 0.472 0.045 0.014 0.480 0.874 0.474 0.190
rep5  0.994 0.504 0.947 0.994 0.829 0.992 0.898 0.927
rep6  0.020 0.553 0.046 0.360 0.028 0.041 0.040 0.125
rep7  0.030 0.417 0.012 0.798 0.325 0.089 0.436 0.010
rep8  0.995 0.688 0.997 0.929 0.055 0.487 0.693 0.088
rep9  0.612 0.040 0.099 0.542 0.436 0.322 0.038 0.257
rep10 0.328 0.286 0.102 0.040 0.963 0.931 0.971 0.089</code></pre>
<pre class="r"><code>### the number of significant genes by comparing each infected condition to the control (lfsr &lt; 0.05)
m.out$num_effects</code></pre>
<pre><code>      cond1 cond2 cond3 cond4 cond5 cond6 cond7 cond8
rep1     19     1    15   206    34    11   250   306
rep2    282   138    65    18     6  1303   148   738
rep3   3831  1065   386    80    29    10    20    39
rep4     25    24   594  2846    37     6    25   358
rep5      0   117    38     0    30     4    13     3
rep6   1479     4   878     4  1257  1926   368   116
rep7   1178   166  3408    57   195   932   120  3828
rep8      1    66     0    13   520   368    45   115
rep9     34   470   230     4    27   209  1020   215
rep10    19   154    25   692    30     4     6    99</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.2 (2019-12-12)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS  10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] edgeR_3.28.1    limma_3.42.2    ashr_2.2-50     workflowr_1.6.2

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.6       knitr_1.30       whisker_0.4      magrittr_2.0.1  
 [5] SQUAREM_2021.1   lattice_0.20-41  R6_2.5.0         rlang_0.4.10    
 [9] stringr_1.4.0    tools_3.6.2      grid_3.6.2       xfun_0.17       
[13] irlba_2.3.3      invgamma_1.1     git2r_0.27.1     htmltools_0.4.0 
[17] assertthat_0.2.1 yaml_2.2.1       digest_0.6.27    rprojroot_2.0.2 
[21] mixsqp_0.3-43    Matrix_1.2-18    later_1.0.0      promises_1.1.0  
[25] fs_1.4.1         glue_1.4.2       evaluate_0.14    rmarkdown_2.3   
[29] stringi_1.4.6    compiler_3.6.2   locfit_1.5-9.4   truncnorm_1.0-8 
[33] httpuv_1.5.2    </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
