<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Yusha Liu" />


<title>Testing differential expression on simulated bulk rna-seq data under the null (same permutation)</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">tbdata_sim</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/YushaLiu/tbdata_sim">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Testing differential expression on simulated bulk rna-seq data under the null (same permutation)</h1>
<h4 class="author">Yusha Liu</h4>
<h4 class="date">2021-3-1</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-03-02
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>tbdata_sim/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20210202code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20210202)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20210202code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20210202)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomYushaLiutbdatasimtree8897e57f5cbb8bd314fa0797af307026832860d6targetblank8897e57a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/YushaLiu/tbdata_sim/tree/8897e57f5cbb8bd314fa0797af307026832860d6" target="_blank">8897e57</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomYushaLiutbdatasimtree8897e57f5cbb8bd314fa0797af307026832860d6targetblank8897e57a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/YushaLiu/tbdata_sim/tree/8897e57f5cbb8bd314fa0797af307026832860d6" target="_blank">8897e57</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    output/.DS_Store

Untracked files:
    Untracked:  code/code_sim_null_summary.R
    Untracked:  code/code_sim_null_v2_summary.R
    Untracked:  output/null_output/
    Untracked:  output/null_v2_output/
    Untracked:  output/sim_null_limma_mash_results.Rds
    Untracked:  output/sim_null_mash_ruv_results.Rds
    Untracked:  output/sim_null_mouthwash_K_3_results.Rds
    Untracked:  output/sim_null_mouthwash_K_3_t_lik_results.Rds
    Untracked:  output/sim_null_mouthwash_K_6_results.Rds
    Untracked:  output/sim_null_mouthwash_K_6_t_lik_results.Rds
    Untracked:  output/sim_null_pois_mash_ruv_results.Rds
    Untracked:  output/sim_null_sva_limma_mash_results.Rds
    Untracked:  output/sim_null_v2_limma_mash_results.Rds
    Untracked:  output/sim_null_v2_mash_ruv_results.Rds
    Untracked:  output/sim_null_v2_mouthwash_K_3_results.Rds
    Untracked:  output/sim_null_v2_mouthwash_K_3_t_lik_results.Rds
    Untracked:  output/sim_null_v2_pois_mash_ruv_results.Rds

Unstaged changes:
    Deleted:    analysis/sim_null_ash.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/sim_null.Rmd</code>) and HTML (<code>docs/sim_null.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/YushaLiu/tbdata_sim/09956f360f082380160c3957c929bb1c590dba82/docs/sim_null.html" target="_blank">09956f3</a>
</td>
<td>
yushaliu
</td>
<td>
2021-03-02
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/YushaLiu/tbdata_sim/blob/49f553ce61548f8f6f6c6ff4a78057a4ca03850d/analysis/sim_null.Rmd" target="_blank">49f553c</a>
</td>
<td>
yushaliu
</td>
<td>
2021-03-02
</td>
<td>
update sim results
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/YushaLiu/tbdata_sim/65693142139dc76ef544a147e3b2ee53f297b068/docs/sim_null.html" target="_blank">6569314</a>
</td>
<td>
yushaliu
</td>
<td>
2021-02-18
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/YushaLiu/tbdata_sim/blob/f89afa3be1d196935c7a100e01656714d4805eb9/analysis/sim_null.Rmd" target="_blank">f89afa3</a>
</td>
<td>
yushaliu
</td>
<td>
2021-02-18
</td>
<td>
update the pairwise comparison results with calibrated shat
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/YushaLiu/tbdata_sim/d3f56325797402e77eb33a6a025120cec9b19479/docs/sim_null.html" target="_blank">d3f5632</a>
</td>
<td>
yushaliu
</td>
<td>
2021-02-03
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/YushaLiu/tbdata_sim/blob/dd01471aca6badef511c4b281e9535560cbf424c/analysis/sim_null.Rmd" target="_blank">dd01471</a>
</td>
<td>
yushaliu
</td>
<td>
2021-02-03
</td>
<td>
Update some analyses
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/YushaLiu/tbdata_sim/3066607a7b57887209431dfad1dcabf3bd4b226a/docs/sim_null.html" target="_blank">3066607</a>
</td>
<td>
yushaliu
</td>
<td>
2021-02-03
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/YushaLiu/tbdata_sim/blob/f4cf8b833534f829734c9cffe37c68995c112de8/analysis/sim_null.Rmd" target="_blank">f4cf8b8</a>
</td>
<td>
yushaliu
</td>
<td>
2021-02-03
</td>
<td>
Add the simulation results
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="simulation-procedure" class="section level2">
<h2>Simulation procedure</h2>
<p>We apply various methods to detect differential exression on bulk rna-seq data simulated under the null. More specifically, we start with a subset of the <a href="https://pubmed.ncbi.nlm.nih.gov/26586179/">tbdata</a> which includes the gene expression data of 12,728 genes from 54 samples infected by 8 different types of bacteria and one non-infected control (6 individuals per condition), measured at 4 hours post-infection. Then we randomly permute the samples with respect to the condition labels so that there no longer exists association between gene expression levels and conditions.</p>
<p>There are two different ways to randomly permute the data: 1) the same permutation to all the genes, so the correlation among genes is preserved on the permuted dataset; 2) independent and different permutation to each gene. Here we simulate null data by applying the same permutation to all the genes; we simulate 10 replicate datasets in total.</p>
</div>
<div id="methods" class="section level2">
<h2>Methods</h2>
<p>We consider the following methods to compare gene expression levels under each infected condition to the control on the simulated null data. For methods related to mash, we always assume <span class="math inline">\(\alpha = 1\)</span>, i.e., the EZ model.</p>
<p>1. limma + mash with common baseline, with or without voom procedure.</p>
<p>2. sva + limma + mash with common baseline.</p>
<p>3. mash + sample-specific random effects without accounting for unwanted variation.</p>
<p>4. mash + sample-specific random effects + ruv, with variance inflation factor constrained to be at least 1. Here we try different number of hidden confounders in the ruv step, and also consider including (i) canonical prior covariances only, (ii) both data-driven and canonical prior covariances. For (ii), to identify the subset of “strong” genes that are used to estimate data-driven covariances, we either (a) directly use <span class="math inline">\(\hat{b}\)</span> and <span class="math inline">\(\hat{s}\)</span> to calculate Z-scores, or (b) first subtract the estimated bias caused by unwanted variation from <span class="math inline">\(\hat{b}\)</span> to get <span class="math inline">\(\hat{\tilde{b}}\)</span>, and then calculate Z-scores using <span class="math inline">\(\hat{\tilde{b}}\)</span> and <span class="math inline">\(\hat{s}\)</span>. The genes with <span class="math inline">\(\max_{r} |Z_{jr}| &gt; 3.5\)</span> are identified as “strong” genes.</p>
<p>5. poisson mash + sample-specific random effects + ruv. We also consider including (i) canonical prior covariances only, (ii) both data-driven and canonical prior covariances.</p>
</div>
<div id="look-at-distributions-of-z-scores-by-comparing-to-the-mean" class="section level2">
<h2>Look at distributions of Z-scores by comparing to the mean</h2>
<pre class="r"><code>library(mashr)
library(edgeR)
library(limma)

### load in the design matrix
design &lt;- readRDS(&quot;../simulations/design.Rds&quot;)
trts &lt;- colnames(design)
R &lt;- length(trts)

### plot the histograms and qqplots of Z-scores across all genes and conditions
par(mfrow=c(2,2))

for(idx in 1:10){
  ### read in data
  counts &lt;- readRDS(paste0(&quot;../simulations/null_data/raw_data&quot;, idx, &quot;.Rds&quot;))
  
  ### normalize and transform the count data
  y &lt;- DGEList(counts)
  y &lt;- calcNormFactors(y)
  dat_cpm &lt;- cpm(y, log = TRUE)
  
  ### extract Bhat and Shat using limma  
  cov_of_interest = 1:ncol(design)
  lmout = limma::lmFit(object = dat_cpm, design = design)
  eout  = limma::eBayes(lmout)
  bhat = lmout$coefficients[,cov_of_interest,drop=FALSE]
  shat = lmout$stdev.unscaled[,cov_of_interest,drop=FALSE] * sqrt(eout$s2.post)
  rownames(shat) = rownames(bhat)
  
  ### compute Z-scores by comparing to the overall mean across all conditions
  mash.data = mash_set_data(bhat, shat, alpha = 1)
  mash.data.L = mash_update_data(mash.data, ref = &quot;mean&quot;)
  mash.Z = mash.data.L$Bhat/mash.data.L$Shat
  
  ### plot the histograms of Z-scores of all genes and conditions
  hist(mash.Z, xlab=&quot;Z-scores&quot;, main=paste0(&quot;Hist of Z-scores \n by comparing to the mean \n across genes and conds for rep=&quot;, idx))
  
  ### plot the Q-Q plot of Z-scores of all genes and conditions
  qqnorm(as.numeric(mash.Z), main=paste0(&quot;Normal qqplot of Z-scores\nby comparing to the mean \n across genes and conds for rep=&quot;, idx))
  qqline(as.numeric(mash.Z))
}</code></pre>
<p><img src="figure/sim_null.Rmd/unnamed-chunk-1-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-1-1">
Past versions of unnamed-chunk-1-1.png
</button>
</p>
<div id="fig-unnamed-chunk-1-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/YushaLiu/tbdata_sim/blob/65693142139dc76ef544a147e3b2ee53f297b068/docs/figure/sim_null.Rmd/unnamed-chunk-1-1.png" target="_blank">6569314</a>
</td>
<td>
yushaliu
</td>
<td>
2021-02-18
</td>
</tr>
<tr>
<td>
<a href="https://github.com/YushaLiu/tbdata_sim/blob/d3f56325797402e77eb33a6a025120cec9b19479/docs/figure/sim_null.Rmd/unnamed-chunk-1-1.png" target="_blank">d3f5632</a>
</td>
<td>
yushaliu
</td>
<td>
2021-02-03
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/sim_null.Rmd/unnamed-chunk-1-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-1-2">
Past versions of unnamed-chunk-1-2.png
</button>
</p>
<div id="fig-unnamed-chunk-1-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/YushaLiu/tbdata_sim/blob/65693142139dc76ef544a147e3b2ee53f297b068/docs/figure/sim_null.Rmd/unnamed-chunk-1-2.png" target="_blank">6569314</a>
</td>
<td>
yushaliu
</td>
<td>
2021-02-18
</td>
</tr>
<tr>
<td>
<a href="https://github.com/YushaLiu/tbdata_sim/blob/d3f56325797402e77eb33a6a025120cec9b19479/docs/figure/sim_null.Rmd/unnamed-chunk-1-2.png" target="_blank">d3f5632</a>
</td>
<td>
yushaliu
</td>
<td>
2021-02-03
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/sim_null.Rmd/unnamed-chunk-1-3.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-1-3">
Past versions of unnamed-chunk-1-3.png
</button>
</p>
<div id="fig-unnamed-chunk-1-3" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/YushaLiu/tbdata_sim/blob/65693142139dc76ef544a147e3b2ee53f297b068/docs/figure/sim_null.Rmd/unnamed-chunk-1-3.png" target="_blank">6569314</a>
</td>
<td>
yushaliu
</td>
<td>
2021-02-18
</td>
</tr>
<tr>
<td>
<a href="https://github.com/YushaLiu/tbdata_sim/blob/d3f56325797402e77eb33a6a025120cec9b19479/docs/figure/sim_null.Rmd/unnamed-chunk-1-3.png" target="_blank">d3f5632</a>
</td>
<td>
yushaliu
</td>
<td>
2021-02-03
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/sim_null.Rmd/unnamed-chunk-1-4.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-1-4">
Past versions of unnamed-chunk-1-4.png
</button>
</p>
<div id="fig-unnamed-chunk-1-4" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/YushaLiu/tbdata_sim/blob/65693142139dc76ef544a147e3b2ee53f297b068/docs/figure/sim_null.Rmd/unnamed-chunk-1-4.png" target="_blank">6569314</a>
</td>
<td>
yushaliu
</td>
<td>
2021-02-18
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/sim_null.Rmd/unnamed-chunk-1-5.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-1-5">
Past versions of unnamed-chunk-1-5.png
</button>
</p>
<div id="fig-unnamed-chunk-1-5" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/YushaLiu/tbdata_sim/blob/65693142139dc76ef544a147e3b2ee53f297b068/docs/figure/sim_null.Rmd/unnamed-chunk-1-5.png" target="_blank">6569314</a>
</td>
<td>
yushaliu
</td>
<td>
2021-02-18
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="result-summary" class="section level2">
<h2>Result summary</h2>
<pre class="r"><code>### limma + mash with common baseline
limma_mash_num_effects &lt;- readRDS(&quot;output/sim_null_limma_mash_results.Rds&quot;)

### number of false positives
limma_mash_num_effects</code></pre>
<pre><code>                rep1 rep2 rep3 rep4 rep5 rep6 rep7 rep8 rep9 rep10
limma_mash         6 3958    5 1856 3073  952 6158 1285 2937  4744
voom_limma_mash 3293 6917 4419 4598 6310 4025 9663 5124 6422  6365</code></pre>
<pre class="r"><code>### false positive rates
round(limma_mash_num_effects/12728, 3)</code></pre>
<pre><code>                 rep1  rep2  rep3  rep4  rep5  rep6  rep7  rep8  rep9 rep10
limma_mash      0.000 0.311 0.000 0.146 0.241 0.075 0.484 0.101 0.231 0.373
voom_limma_mash 0.259 0.543 0.347 0.361 0.496 0.316 0.759 0.403 0.505 0.500</code></pre>
<pre class="r"><code>### sva + limma + mash with common baseline, with different numbers of hidden confounders in sva
sva_limma_mash_num_effects &lt;- readRDS(&quot;output/sim_null_sva_limma_mash_results.Rds&quot;)

### number of false positives
sva_limma_mash_num_effects</code></pre>
<pre><code>      rep1  rep2  rep3  rep4  rep5  rep6  rep7  rep8  rep9 rep10
K=5   7138  8457  7535  7959  6685  6908  9158  8062  7452  8539
K=15  9647 10295  9525 10338  9469 10475 10084 10529  9821 10128
K=20 10528 10834 10273 11104 10057 10582 10227 11003 10380 10743</code></pre>
<pre class="r"><code>### false positive rates
round(sva_limma_mash_num_effects/12728, 3)</code></pre>
<pre><code>      rep1  rep2  rep3  rep4  rep5  rep6  rep7  rep8  rep9 rep10
K=5  0.561 0.664 0.592 0.625 0.525 0.543 0.720 0.633 0.585 0.671
K=15 0.758 0.809 0.748 0.812 0.744 0.823 0.792 0.827 0.772 0.796
K=20 0.827 0.851 0.807 0.872 0.790 0.831 0.804 0.864 0.816 0.844</code></pre>
<pre class="r"><code>### mash + sample-specific random effects, with or without ruv
mash_ruv_num_effects &lt;- readRDS(&quot;output/sim_null_mash_ruv_results.Rds&quot;)

### number of false positives
mash_ruv_num_effects</code></pre>
<pre><code>                rep1 rep2 rep3 rep4 rep5 rep6 rep7 rep8 rep9 rep10
mash_no_ruv        2 5002  120 1311 1781  693 4563 1655 2960  4901
mash_ruv_K_5_c     0    0    0    0    0    0    0  137    0     0
mash_ruv_K_5_o   959 3089 1376 3165 1578 1758 5878 4679 2214  2895
mash_ruv_K_5_w   630 1185  455 2856 1159 1149 5219 3745 2025  2101
mash_ruv_K_15_c    0   32    0    0    0  139    0   87    0     0
mash_ruv_K_15_o  512 1175  868 1372   78 2152 2388 2671 1717    82
mash_ruv_K_15_w  307 1099  164  818    8 2233 2120 2456 1508    44
mash_ruv_K_20_c    0  254    0    0    0  245    0    0    0     0
mash_ruv_K_20_o  252 1448  190  432    0 1851 1714 1235 1600   323
mash_ruv_K_20_w   67 1348   18  298    0 1712 1343  837 1290   179</code></pre>
<pre class="r"><code>### false positive rates
round(mash_ruv_num_effects/12728, 3)</code></pre>
<pre><code>                 rep1  rep2  rep3  rep4  rep5  rep6  rep7  rep8  rep9 rep10
mash_no_ruv     0.000 0.393 0.009 0.103 0.140 0.054 0.359 0.130 0.233 0.385
mash_ruv_K_5_c  0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.011 0.000 0.000
mash_ruv_K_5_o  0.075 0.243 0.108 0.249 0.124 0.138 0.462 0.368 0.174 0.227
mash_ruv_K_5_w  0.049 0.093 0.036 0.224 0.091 0.090 0.410 0.294 0.159 0.165
mash_ruv_K_15_c 0.000 0.003 0.000 0.000 0.000 0.011 0.000 0.007 0.000 0.000
mash_ruv_K_15_o 0.040 0.092 0.068 0.108 0.006 0.169 0.188 0.210 0.135 0.006
mash_ruv_K_15_w 0.024 0.086 0.013 0.064 0.001 0.175 0.167 0.193 0.118 0.003
mash_ruv_K_20_c 0.000 0.020 0.000 0.000 0.000 0.019 0.000 0.000 0.000 0.000
mash_ruv_K_20_o 0.020 0.114 0.015 0.034 0.000 0.145 0.135 0.097 0.126 0.025
mash_ruv_K_20_w 0.005 0.106 0.001 0.023 0.000 0.135 0.106 0.066 0.101 0.014</code></pre>
<pre class="r"><code>### poisson mash + sample-specific random effects + ruv
pois_mash_num_effects &lt;- readRDS(&quot;output/sim_null_pois_mash_ruv_results.Rds&quot;)

### number of false positives
pois_mash_num_effects</code></pre>
<pre><code>                     rep1 rep2 rep3 rep4 rep5 rep6 rep7 rep8 rep9 rep10
pois_mash_ruv_K_15_c 1460 3173 2010  193 3057  836  113  560  343  4970
pois_mash_ruv_K_15_w 4627 5139 4395 3312 6246 7890   NA 2440 8480    NA</code></pre>
<pre class="r"><code>### false positive rates
round(pois_mash_num_effects/12728, 3)</code></pre>
<pre><code>                      rep1  rep2  rep3  rep4  rep5  rep6  rep7  rep8  rep9
pois_mash_ruv_K_15_c 0.115 0.249 0.158 0.015 0.240 0.066 0.009 0.044 0.027
pois_mash_ruv_K_15_w 0.364 0.404 0.345 0.260 0.491 0.620    NA 0.192 0.666
                     rep10
pois_mash_ruv_K_15_c  0.39
pois_mash_ruv_K_15_w    NA</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.2 (2019-12-12)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS  10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] edgeR_3.28.1    limma_3.42.2    mashr_0.2.21    ashr_2.2-50    
[5] workflowr_1.6.2

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.6       compiler_3.6.2   later_1.0.0      git2r_0.27.1    
 [5] plyr_1.8.6       tools_3.6.2      digest_0.6.27    evaluate_0.14   
 [9] lattice_0.20-41  rlang_0.4.10     Matrix_1.2-18    yaml_2.2.1      
[13] mvtnorm_1.1-0    xfun_0.17        invgamma_1.1     stringr_1.4.0   
[17] knitr_1.30       fs_1.4.1         locfit_1.5-9.4   rprojroot_2.0.2 
[21] grid_3.6.2       glue_1.4.2       R6_2.5.0         rmarkdown_2.3   
[25] mixsqp_0.3-43    irlba_2.3.3      rmeta_3.0        magrittr_2.0.1  
[29] whisker_0.4      promises_1.1.0   htmltools_0.4.0  abind_1.4-5     
[33] assertthat_0.2.1 httpuv_1.5.2     stringi_1.4.6    truncnorm_1.0-8 
[37] SQUAREM_2021.1  </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
