---
title: "Testing differential expression on simulated bulk rna-seq data under the null (same permutation)"
author: "Yusha Liu"
header-includes:
date: "2021-3-1"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Simulation procedure
We apply various methods to detect differential exression on bulk rna-seq data simulated under the null. More specifically, we start with a subset of the [tbdata](https://pubmed.ncbi.nlm.nih.gov/26586179/) which includes the gene expression data of 12,728 genes from 54 samples infected by 8 different types of bacteria and one non-infected control (6 individuals per condition), measured at 4 hours post-infection. Then we randomly permute the samples with respect to the condition labels so that there no longer exists association between gene expression levels and conditions. 

There are two different ways to randomly permute the data: 1) the same permutation to all the genes, so the correlation among genes is preserved on the permuted dataset; 2) independent and different permutation to each gene. Here we simulate null data by applying the same permutation to all the genes; we simulate 10 replicate datasets in total.


## Methods
We apply ash and mouthwash to do pairwise comparisons of gene expressions between each infected condition and the control on the simulated null data.

We implement the following 4 versions of ash:

\noindent 1. $\alpha=1$ (EZ model), original shat.

\noindent 2. $\alpha=1$ (EZ model), shat calibrated based on Z-score.

\noindent 3. $\alpha=0$ (EE model), original shat.

\noindent 4. $\alpha=0$ (EE model), shat calibrated based on Z-score.

We implement the following 4 versions of mouthwash, while setting $\alpha=1$ (EZ model) in all cases:

\noindent 1. Number of hidden confounders $K=3$, original shat.

\noindent 2. Number of hidden confounders $K=3$, shat calibrated based on Z-score.

\noindent 3. Number of hidden confounders $K=6$, original shat.

\noindent 4. Number of hidden confounders $K=6$, shat calibrated based on Z-score.



## Load in the packages and the shat calibration function 
```{r warning=FALSE, message=FALSE}
library(ashr)
library(edgeR)
library(limma)

### the function to calibrate shat under t likelihood based on Z score
adjust_by_t <- function(betahat, sebetahat, df) {
  ## Check input -----------------------------------------
  assertthat::are_equal(length(betahat), length(sebetahat))
  assertthat::assert_that(length(betahat) == length(df) | length(df) == 1)
  assertthat::assert_that(df > 0)
  
  ## Convert t to z --------------------------------------
  zstats <- stats::qnorm(stats::pt(q = betahat / sebetahat, df = df))
  snew <- c(betahat / zstats)
  
  return(snew)
}


### load in the design matrix
design <- readRDS("../simulations/design.Rds")
trts <- colnames(design)
R <- length(trts)

```


## Ash:  $\alpha=1$ (EZ model), original shat.
```{r warning=FALSE, message=FALSE}
num_effects1 <- matrix(NA, nrow=10, ncol=R-1)

for(idx in 1:10){
  scdata <- readRDS(paste0("../simulations/null_data/raw_data", idx, ".Rds"))
  
  ### iterate over conditions 
  for(r in 1:8){
    ### take data from one infected condition and the control
    counts <- scdata[, (design[,1]==1 | design[,r+1]==1)]
    
    ### normalize and transform the count data
    y <- DGEList(counts)
    y <- calcNormFactors(y)
    dat_cpm <- cpm(y, log = TRUE)
    
    ### construct the covariate matrix
    data.X <- model.matrix(~design[(design[,1]==1 | design[,r+1]==1), r+1])
    colnames(data.X) <- c("Intercept", "Trt")
    
    ### compute bhat and shat (of the difference between the given condition and the negative control)
    cov_of_interest = 2
    lmout = limma::lmFit(object = dat_cpm, design = data.X)
    eout = limma::eBayes(lmout)
    bhat = lmout$coefficients[,cov_of_interest]
    shat = lmout$stdev.unscaled[,cov_of_interest] * sqrt(eout$s2.post)
    
    ### apply ash
    ash.out = ash(betahat=bhat, sebetahat=shat, mixcompdist="normal", alpha=1, lik=lik_normal())
    num_effects1[idx,r] = sum(ash.out$result$lfsr < 0.05)
  }
}


### the number of significant genes by comparing each infected condition to the control (lfsr < 0.05)
num_effects1

```


## Ash: $\alpha=1$ (EZ model), shat calibrated based on Z-score.
```{r warning=FALSE, message=FALSE}
num_effects2 <- matrix(NA, nrow=10, ncol=R-1)

for(idx in 1:10){
  scdata <- readRDS(paste0("../simulations/null_data/raw_data", idx, ".Rds"))
  
  ### iterate over conditions 
  for(r in 1:8){
    ### take data from one infected condition and the control
    counts <- scdata[, (design[,1]==1 | design[,r+1]==1)]
    
    ### normalize and transform the count data
    y <- DGEList(counts)
    y <- calcNormFactors(y)
    dat_cpm <- cpm(y, log = TRUE)
    
    ### construct the covariate matrix
    data.X <- model.matrix(~design[(design[,1]==1 | design[,r+1]==1), r+1])
    colnames(data.X) <- c("Intercept", "Trt")
    
    ### compute bhat and shat (of the difference between the given condition and the negative control)
    cov_of_interest = 2
    lmout = limma::lmFit(object = dat_cpm, design = data.X)
    eout = limma::eBayes(lmout)
    bhat = lmout$coefficients[,cov_of_interest]
    shat = lmout$stdev.unscaled[,cov_of_interest] * sqrt(eout$s2.post)
    shat.adj = adjust_by_t(bhat, shat, df=ncol(dat_cpm)-2)
    
    ### apply ash
    ash.out = ash(betahat=bhat, sebetahat=shat.adj, mixcompdist="normal", alpha=1, lik=lik_normal())
    num_effects2[idx,r] = sum(ash.out$result$lfsr < 0.05)
  }
}

### the number of significant genes by comparing each infected condition to the control (lfsr < 0.05)
num_effects2

```


## Ash: $\alpha=0$ (EE model), original shat.
```{r warning=FALSE, message=FALSE}
num_effects3 <- matrix(NA, nrow=10, ncol=R-1)

for(idx in 1:10){
  scdata <- readRDS(paste0("../simulations/null_data/raw_data", idx, ".Rds"))
  
  ### iterate over conditions 
  for(r in 1:8){
    ### take data from one infected condition and the control
    counts <- scdata[, (design[,1]==1 | design[,r+1]==1)]
    
    ### normalize and transform the count data
    y <- DGEList(counts)
    y <- calcNormFactors(y)
    dat_cpm <- cpm(y, log = TRUE)
    
    ### construct the covariate matrix
    data.X <- model.matrix(~design[(design[,1]==1 | design[,r+1]==1), r+1])
    colnames(data.X) <- c("Intercept", "Trt")
    
    ### compute bhat and shat (for the difference between the given condition and the negative control)
    cov_of_interest = 2
    lmout = limma::lmFit(object = dat_cpm, design = data.X)
    eout = limma::eBayes(lmout)
    bhat = lmout$coefficients[,cov_of_interest]
    shat = lmout$stdev.unscaled[,cov_of_interest] * sqrt(eout$s2.post)
    
    ### apply ash
    ash.out = ash(betahat=bhat, sebetahat=shat, mixcompdist="normal", alpha=0, lik=lik_normal())
    num_effects3[idx,r] = sum(ash.out$result$lfsr < 0.05)
  }
}

### the number of significant genes by comparing each infected condition to the control (lfsr < 0.05)
num_effects3

```


## Ash: $\alpha=0$ (EE model), shat calibrated based on Z-score.
```{r warning=FALSE, message=FALSE}
num_effects4 <- matrix(NA, nrow=10, ncol=R-1)

for(idx in 1:10){
  scdata <- readRDS(paste0("../simulations/null_data/raw_data", idx, ".Rds"))
  
  ### iterate over conditions 
  for(r in 1:8){
    ### take data from one infected condition and the control
    counts <- scdata[, (design[,1]==1 | design[,r+1]==1)]
    
    ### normalize and transform the count data
    y <- DGEList(counts)
    y <- calcNormFactors(y)
    dat_cpm <- cpm(y, log = TRUE)
    
    ### construct the covariate matrix
    data.X <- model.matrix(~design[(design[,1]==1 | design[,r+1]==1), r+1])
    colnames(data.X) <- c("Intercept", "Trt")
    
    ### compute bhat and shat (for the difference between the given condition and the negative control)
    cov_of_interest = 2
    lmout = limma::lmFit(object = dat_cpm, design = data.X)
    eout = limma::eBayes(lmout)
    bhat = lmout$coefficients[,cov_of_interest]
    shat = lmout$stdev.unscaled[,cov_of_interest] * sqrt(eout$s2.post)
    shat.adj = adjust_by_t(bhat, shat, df=ncol(dat_cpm)-2)
    
    ### apply ash
    ash.out = ash(betahat=bhat, sebetahat=shat.adj, mixcompdist="normal", alpha=0, lik=lik_normal())
    num_effects4[idx,r] = sum(ash.out$result$lfsr < 0.05)
  }
}

### the number of significant genes by comparing each infected condition to the control (lfsr < 0.05)
num_effects4

```


## Mouthwash: number of hidden confounders $K=3$, original shat.
```{r warning=FALSE, message=FALSE}
m.out <- readRDS("output/sim_null_mouthwash_K_3_results.Rds")

### the estimate of the proportion of null genes
round(m.out$pi0, 3)

### the number of significant genes by comparing each infected condition to the control (lfsr < 0.05)
m.out$num_effects

```

## Mouthwash: number of hidden confounders $K=3$, shat calibrated based on Z-score.
```{r warning=FALSE, message=FALSE}
m.out <- readRDS("output/sim_null_mouthwash_K_3_t_lik_results.Rds")

### the estimate of the proportion of null genes
round(m.out$pi0, 3)

### the number of significant genes by comparing each infected condition to the control (lfsr < 0.05)
m.out$num_effects

```

## Mouthwash: number of hidden confounders $K=6$, original shat.
```{r warning=FALSE, message=FALSE}
m.out <- readRDS("output/sim_null_mouthwash_K_6_results.Rds")

### the estimate of the proportion of null genes
round(m.out$pi0, 3)

### the number of significant genes by comparing each infected condition to the control (lfsr < 0.05)
m.out$num_effects

```

## Mouthwash: number of hidden confounders $K=6$, shat calibrated based on Z-score.
```{r warning=FALSE, message=FALSE}
m.out <- readRDS("output/sim_null_mouthwash_K_6_t_lik_results.Rds")

### the estimate of the proportion of null genes
round(m.out$pi0, 3)

### the number of significant genes by comparing each infected condition to the control (lfsr < 0.05)
m.out$num_effects

```