---
title: "Testing differential expression on simulated bulk rna-seq data under the null (same permutation)"
author: "Yusha Liu"
header-includes:
date: "2021-3-1"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Simulation procedure
We apply various methods to detect differential exression on bulk rna-seq data simulated under the null. More specifically, we start with a subset of the [tbdata](https://pubmed.ncbi.nlm.nih.gov/26586179/) which includes the gene expression data of 12,728 genes from 54 samples infected by 8 different types of bacteria and one non-infected control (6 individuals per condition), measured at 4 hours post-infection. Then we randomly permute the samples with respect to the condition labels so that there no longer exists association between gene expression levels and conditions. 

There are two different ways to randomly permute the data: 1) the same permutation to all the genes, so the correlation among genes is preserved on the permuted dataset; 2) independent and different permutation to each gene. Here we simulate null data by applying the same permutation to all the genes; we simulate 10 replicate datasets in total.


## Methods
We consider the following methods to compare gene expression levels under each infected condition to the control on the simulated null data. For methods related to mash, we always assume $\alpha = 1$, i.e., the EZ model. 

\noindent 1. limma + mash with common baseline, with or without voom procedure.

\noindent 2. sva + limma + mash with common baseline.

\noindent 3. mash + sample-specific random effects without accounting for unwanted variation.

\noindent 4. mash + sample-specific random effects + ruv, with variance inflation factor constrained to be at least 1. Here we try different number of hidden confounders in the ruv step, and also consider including (i) canonical prior covariances only, (ii) both data-driven and canonical prior covariances. For (ii), to identify the subset of "strong" genes that are used to estimate data-driven covariances, we either (a) directly use $\hat{b}$ and $\hat{s}$ to calculate Z-scores, or (b) first subtract the estimated bias caused by unwanted variation from $\hat{b}$ to get $\hat{\tilde{b}}$, and then calculate Z-scores using $\hat{\tilde{b}}$ and $\hat{s}$. The genes with $\max_{r} |Z_{jr}| > 3.5$ are identified as "strong" genes.

\noindent 5. poisson mash + sample-specific random effects + ruv. We also consider including (i) canonical prior covariances only, (ii) both data-driven and canonical prior covariances.


## Look at distributions of Z-scores by comparing to the mean
```{r warning=FALSE, message=FALSE}
library(mashr)
library(edgeR)
library(limma)

### load in the design matrix
design <- readRDS("../simulations/design.Rds")
trts <- colnames(design)
R <- length(trts)

### plot the histograms and qqplots of Z-scores across all genes and conditions
par(mfrow=c(2,2))

for(idx in 1:10){
  ### read in data
  counts <- readRDS(paste0("../simulations/null_data/raw_data", idx, ".Rds"))
  
  ### normalize and transform the count data
  y <- DGEList(counts)
  y <- calcNormFactors(y)
  dat_cpm <- cpm(y, log = TRUE)
  
  ### extract Bhat and Shat using limma  
  cov_of_interest = 1:ncol(design)
  lmout = limma::lmFit(object = dat_cpm, design = design)
  eout  = limma::eBayes(lmout)
  bhat = lmout$coefficients[,cov_of_interest,drop=FALSE]
  shat = lmout$stdev.unscaled[,cov_of_interest,drop=FALSE] * sqrt(eout$s2.post)
  rownames(shat) = rownames(bhat)
  
  ### compute Z-scores by comparing to the overall mean across all conditions
  mash.data = mash_set_data(bhat, shat, alpha = 1)
  mash.data.L = mash_update_data(mash.data, ref = "mean")
  mash.Z = mash.data.L$Bhat/mash.data.L$Shat
  
  ### plot the histograms of Z-scores of all genes and conditions
  hist(mash.Z, xlab="Z-scores", main=paste0("Hist of Z-scores \n by comparing to the mean \n across genes and conds for rep=", idx))
  
  ### plot the Q-Q plot of Z-scores of all genes and conditions
  qqnorm(as.numeric(mash.Z), main=paste0("Normal qqplot of Z-scores\nby comparing to the mean \n across genes and conds for rep=", idx))
  qqline(as.numeric(mash.Z))
}

```


## Result summary
```{r warning=FALSE, message=FALSE}
### limma + mash with common baseline
limma_mash_num_effects <- readRDS("output/sim_null_limma_mash_results.Rds")

### number of false positives
limma_mash_num_effects

### false positive rates
round(limma_mash_num_effects/12728, 3)

```


```{r warning=FALSE, message=FALSE}
### sva + limma + mash with common baseline, with different numbers of hidden confounders in sva
sva_limma_mash_num_effects <- readRDS("output/sim_null_sva_limma_mash_results.Rds")

### number of false positives
sva_limma_mash_num_effects

### false positive rates
round(sva_limma_mash_num_effects/12728, 3)

```


```{r warning=FALSE, message=FALSE}
### mash + sample-specific random effects, with or without ruv
mash_ruv_num_effects <- readRDS("output/sim_null_mash_ruv_results.Rds")

### number of false positives
mash_ruv_num_effects

### false positive rates
round(mash_ruv_num_effects/12728, 3)

```


```{r warning=FALSE, message=FALSE}
### poisson mash + sample-specific random effects + ruv
pois_mash_num_effects <- readRDS("output/sim_null_pois_mash_ruv_results.Rds")

### number of false positives
pois_mash_num_effects

### false positive rates
round(pois_mash_num_effects/12728, 3)

```
