---
title: "Testing differential expression on simulated bulk rna-seq data under the null (different permutation)"
author: "Yusha Liu"
header-includes:
date: "2021-3-1"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Simulation procedure
We apply various methods to detect differential exression on bulk rna-seq data simulated under the null. More specifically, we start with a subset of the [tbdata](https://pubmed.ncbi.nlm.nih.gov/26586179/) which includes the gene expression data of 12,728 genes from 54 samples infected by 8 different types of bacteria and one non-infected control (6 individuals per condition), measured at 4 hours post-infection. Then we randomly permute the samples with respect to the condition labels so that there no longer exists association between gene expression levels and conditions. 

There are two different ways to randomly permute the data: 1) the same permutation to all the genes, so the correlation among genes is preserved on the permuted dataset; 2) independent and different permutation to each gene. Here we simulate null data by applying different and independent permutations to each gene; we simulate 10 replicate datasets in total.


## Methods: multi-condition comparisons
We consider the following methods to compare gene expression levels under each infected condition to the control on the simulated null data. For methods related to mash, we always assume $\alpha = 1$, i.e., the EZ model. 

\noindent 1. limma + mash with common baseline, with or without voom procedure.

\noindent 2. sva + limma + mash with common baseline.

\noindent 3. mash + sample-specific random effects without accounting for unwanted variation.

\noindent 4. mash + sample-specific random effects + ruv, with variance inflation factor constrained to be at least 1. Here we include both data-driven and canonical prior covariances. To identify the subset of "strong" genes that are used to estimate data-driven covariances, we first subtract the estimated bias caused by unwanted variation from $\hat{b}$ to get $\hat{\tilde{b}}$, and then calculate Z-scores using $\hat{\tilde{b}}$ and $\hat{s}$. The genes with $\max_{r} |Z_{jr}| > 3.5$ are identified as "strong" genes.

\noindent 5. poisson mash + sample-specific random effects + ruv. We consider including (i) canonical prior covariances only, and (ii) both data-driven and canonical prior covariances.


## Methods: pairwise comparisons
We apply mouthwash to do pairwise comparisons of gene expressions between each infected condition and the control on the simulated null data.


## Results: limma + mash with common baseline
```{r warning=FALSE, message=FALSE}
limma_mash_num_effects <- readRDS("output/sim_null_v2_limma_mash_results.Rds")

### number of false positives
limma_mash_num_effects

### false positive rates
round(limma_mash_num_effects/12728, 3)

```


## Results: mash + sample-specific random effects, with or without ruv
```{r warning=FALSE, message=FALSE}
mash_ruv_num_effects <- readRDS("output/sim_null_v2_mash_ruv_results.Rds")

### number of false positives
mash_ruv_num_effects

### false positive rates
round(mash_ruv_num_effects/12728, 3)

```


## Results: poisson mash + sample-specific random effects + ruv
```{r warning=FALSE, message=FALSE}
pois_mash_num_effects <- readRDS("output/sim_null_v2_pois_mash_ruv_results.Rds")

### number of false positives
pois_mash_num_effects

### false positive rates
round(pois_mash_num_effects/12728, 3)

```


## Results: pairwise comparisons with control using mouthwash (K=3, sprop=1)
```{r warning=FALSE, message=FALSE}
m.out <- readRDS("output/sim_null_v2_mouthwash_K_3_results.Rds")

### the estimate of the proportion of null genes
round(m.out$pi0, 3)

### the number of significant genes by comparing each infected condition to the control (lfsr < 0.05)
m.out$num_effects

```


## Results: pairwise comparisons with control using mouthwash, with calibrated shat (K=3, sprop=1)
```{r warning=FALSE, message=FALSE}
m.out <- readRDS("output/sim_null_v2_mouthwash_K_3_t_lik_results.Rds")

### the estimate of the proportion of null genes
round(m.out$pi0, 3)

### the number of significant genes by comparing each infected condition to the control (lfsr < 0.05)
m.out$num_effects

```