---
title: "Testing differential expression on simulated bulk rna-seq data under the null (different permutation)"
author: "Yusha Liu"
header-includes:
date: "2021-2-2"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Simulation procedure
We apply various methods to detect differential exression on bulk rna-seq data simulated under the null. More specifically, we start with a subset of the [tbdata](https://pubmed.ncbi.nlm.nih.gov/26586179/) which includes the gene expression data of 12,728 genes from 54 samples infected by 8 different types of bacteria and one non-infected control (6 individuals per condition), measured at 4 hours post-infection. Then we randomly permute the samples with respect to the condition labels so that there no longer exists association between gene expression levels and conditions. 

There are two different ways to randomly permute the data: 1) the same permutation to all the genes, so the correlation among genes is preserved on the permuted dataset; 2) independent and different permutation to each gene. Here we simulate null data by applying different and independent permutations to each gene; we simulate 10 replicate datasets in total.


## Methods
We consider the following methods to compare gene expression levels under each infected condition to the control on the simulated null data. For methods related to mash, we always assume $\alpha = 1$, i.e., the EZ model. 

\noindent 1. limma + mash with common baseline.

\noindent 2. sva + limma + mash with common baseline.

\noindent 3. mash + sample-specific random effects.

\noindent 4. mash + sample-specific random effects + ruv, with variance inflation factor constrained to be at least 1. 

\noindent 5. poisson mash + sample-specific random effects + ruv.


## Result summary
```{r warning=FALSE, message=FALSE}
library(mashr)

### limma + mash with common baseline
num_effects_1 <- rep(NA, 10)

for(i in 1:10){
  mash.fit <- readRDS(paste0("output/null_v2_output/mash_fit_rep_", i, ".Rds"))
  idx.mash <- get_significant_results(mash.fit)
  num_effects_1[i] <- length(idx.mash)
}

### number of false positives
num_effects_1

### false positive rates
round(num_effects_1/nrow(mash.fit$result$lfsr), 3)

```


```{r warning=FALSE, message=FALSE}
### sva + limma + mash with common baseline 
num_effects_2 <- rep(NA, 10)

for(i in 1:10){
  mash.fit <- readRDS(paste0("output/null_v2_output/sva_mash_fit_rep_", i, ".Rds"))
  idx.mash <- get_significant_results(mash.fit)
  num_effects_2[i] <- length(idx.mash)
}

### number of false positives
num_effects_2

### false positive rates
round(num_effects_2/nrow(mash.fit$result$lfsr), 3)

```


```{r warning=FALSE, message=FALSE}
### mash + sample-specific random effects
num_effects_3 <- rep(NA, 10)

for(i in 1:10){
  mash.fit <- readRDS(paste0("output/null_v2_output/mash_no_ruv_posterior_rep_", i, ".Rds"))
  num_effects_3[i] <- sum(apply(mash.fit$lfsr, 1, min) < 0.05)
}

### number of false positives
num_effects_3

### false positive rates
round(num_effects_3/nrow(mash.fit$lfsr), 3)

```


```{r warning=FALSE, message=FALSE}
### mash + sample-specific random effects + ruv
num_effects_4 <- rep(NA, 10)

for(i in 1:10){
  mash.fit <- readRDS(paste0("output/null_v2_output/mash_ruv_posterior_rep_", i, ".Rds"))
  num_effects_4[i] <- sum(apply(mash.fit$lfsr, 1, min) < 0.05)
}

### number of false positives
num_effects_4

### false positive rates
round(num_effects_4/nrow(mash.fit$lfsr), 3)

```



```{r warning=FALSE, message=FALSE}
### poisson mash + sample-specific random effects + ruv
num_effects_5 <- rep(NA, 10)

for(i in 1:10){
  pois.fit <- readRDS(paste0("output/null_v2_output/pois_mash_ruv_posterior_rep_", i, ".Rds"))
  num_effects_5[i] <- sum(apply(pois.fit$lfsr, 1, min) < 0.05)
}

### number of false positives
num_effects_5

### false positive rates
round(num_effects_5/nrow(pois.fit$lfsr), 3)

```


## Look at pairwise comparisons with control using MOUTHWASH (sprop=1)
```{r warning=FALSE, message=FALSE}
m.out <- readRDS("output/mouthwash_sim_null_v2_summary.Rds")

## the estimate of the proportion of null genes
round(m.out$pi0,3)

## the number of DE genes by comparing each infected condition to the control (lfsr < 0.05)
m.out$num_effects_1

## the number of DE genes by comparing each infected condition to the control (lfsr < 0.05/8)
m.out$num_effects_2

```

